### DISCUSSION TITLE
How to use variable in parsed function?

### DISCUSSION THREAD

**Discussion by aiskhak** [idaholab/moose#19363] [Q&A General] [answered]
I want to implement the Boussinesq approximation in a MOOSE-based SAM. TO do so I have the equation of state with constant density, but I want to define body force that depends on temperature:

```
[GlobalParams]
  temperature = T 
  ...
[]

[Functions]
  [./body_f]
    # Boussinesq body force
    type = ParsedFunction
    value = 9.8*0.38*T
    vars = 'T'
    vals = T
  [../]
[]

[Variables]
  [./T]
    initial_condition = 0.0
  [../]
  ...
[]

[Kernels]
  ...
  [./body_force]
    type = BodyForce
    variable = v
    function = body_f
  [../]
```

However, MOOSE does not allow using variables in parsed functions. What would be the solution?

**Comment by ykvishal** [comment, marked as answer]
I think that there are 2 more options [MaskedBodyForce ](https://mooseframework.inl.gov/source/kernels/MaskedBodyForce.html)and [INSADBoussinesqBodyForce](https://mooseframework.inl.gov/source/kernels/INSADBoussinesqBodyForce.html). I hope that one of them will work.

**Reply by aiskhak** [reply]
Thank you for the reply. It seems like MaskedBodyForce allows only using material property, but not a variable. INSADBoussinesqBodyForce works only with NS module, while I have different kernels...

**Reply by GiudGiud** [reply]
It's the right lead though.
Dont use a function, use a kernel directly. Functions cant depend on non-linear variables.

**Reply by aiskhak** [reply]
So you are saying to modify a kernel (requires access to sources)?

**Reply by GiudGiud** [reply]
do you not have access to SAM source code?
I would say add a kernel that does what you want

**Reply by aiskhak** [reply]
yes, I do. Thanks for the advice!

**Reply by travismui** [reply]
@aiskhak You can use a [CoupledForce](https://mooseframework.inl.gov/source/kernels/CoupledForce.html) kernel to add a scalar factor temperature-dependent body force. Note that Moose syntax will work in line with the SAM syntax, so you can achieve this directly from the input. If you have more complicated requirements, please feel free to contact the SAM developers directly to discuss your needs.

**Reply by aiskhak** [reply]
Thanks, this is probably what I need.
